consensus/istanbul/core/roundstate_persist.go

```diff
@@ -0,0 +1,352 @@
+// Copyright 2017 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package core
+
+import (
+	"bytes"
+	"encoding/binary"
+	"math/big"
+	"os"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/consensus/istanbul"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/syndtr/goleveldb/leveldb"
+	lvlerrors "github.com/syndtr/goleveldb/leveldb/errors"
+	"github.com/syndtr/goleveldb/leveldb/opt"
+	"github.com/syndtr/goleveldb/leveldb/storage"
+)
+
+const (
+	dbVersion    = 1
+	dbVersionKey = "version"  // Version of the database to flush if changes
+	lastViewKey  = "lastView" // Last View that we know of
+)
+
+func newRoundStateWithPersistence(sequence *big.Int, validatorSet istanbul.ValidatorSet, proposer istanbul.Validator, path string) (RoundState, error) {
+	logger := log.New("func", "newRoundStateWithPersistence")
+
+	var db *leveldb.DB
+	var err error
+	if path == "" {
+		logger.Info("Open roundstate db", "path", "inmemory")
+		db, err = newMemoryDB()
+	} else {
+		logger.Info("Open roundstate db", "path", path)
+		db, err = newPersistentDB(path)
+	}
+
+	if err != nil {
+		logger.Error("Failed to open roundstate db", "err", err)
+		return nil, err
+	}
+
+	view := &istanbul.View{
+		Sequence: sequence,
+		Round:    common.Big0,
+	}
+
+	lastStoredView, err := getLastStoredView(db)
+
+	if err != nil && err != leveldb.ErrNotFound {
+		logger.Error("Failed to fetch lastStoredView", "err", err)
+		db.Close()
+		return nil, err
+	}
+
+	if err == leveldb.ErrNotFound || lastStoredView.Cmp(view) < 0 {
```

Comment:
I think adding a comment here stating that this function will create a new round state object if the last stored view is old will be helpful.

Reply:
Added a comment and renamed the function to `createOrRestoreRoundState()` so that the name makes it clearer